subsys/net/l2/ppp/ppp_l2.c

```diff
@@ -351,22 +350,30 @@ static void net_ppp_mgmt_evt_handler(struct net_mgmt_event_callback *cb, uint32_
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(PPP)) {
 		return;
 	}
+#if defined(CONFIG_NET_PPP)
+	const struct ppp_api *ppp =
+		net_if_get_device(iface)->api;
+#endif
 
 	ctx = net_if_l2_data(iface);
 
-	if (net_if_is_carrier_ok(iface)) {
```

Comment:
Should be noted that with PPP, there are (unfortunately) still 2 separate carrier on/offs, net if (not really used with ppp afaik) and also PPP carrier on/off (have been and is used). Mentioned PR break/changed functionality of _PPP_ carrier on/off events. Changing/fixing that back more like it was, is one of purposes of this PR, not implementing net_if_carrier_on/off support.
Eventually, when implemented, net if carrier on/off could be like you said: indicating carrier on/off and hopefully we could get rid of ppp carrier on/off. The best would be that with ppp, NET_IF is notified to be usable only after IP is up (if IPCP used). But these are not the topic of this PR.

In my usage this highlighted code piece (condition for net_if_is_carrier_ok()) didn't work correctly as mixing 2 separate carrier on/offs here. Resulting odd PPP carrier events, for example NET_EVENT_IF_DOWN resulted PPP carrier ON.

Reply:
I agree, I left in the event in case it was being used somewhere downstream. This event should be called from a lower level which actually knows if the carrier is up/down, like a modem device driver which knows if the the data channel is ready and modem is registered to the network.

This does not mean that it should be placed in place of the proper events when LCP is running or dropped though right?

Reply:
As already said, this code here is mixing 2 separate issues net_if_carrier_on() is and has been different than PPP carrier ON.
Additionally, as you said, if understood correctly, PR #59124 creates undocumented dependency that needs to be met in order to work currently. And in your case, that "lower level" that sets net_if_carrier_on/off is in your own code, not in Zephyr code, right? At least I cannot find that in gsm_ppp.c

Reply:
The documentation describing how the L2/L1 interacts with the upper network stack is described here [[Network interface state management](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html)

The PR #59124 did not update the gsm_ppp.c driver to implement `net_if_carrier_on()/off()`. It was simply updated to not call the internal enable() function of the network interface directly. `enable()` is actually called by the net_if subsystem here.
https://github.com/zephyrproject-rtos/zephyr/blob/01ba923bb203c749f96bfa9d757f515d1fd91838/subsys/net/ip/net_if.c#L4356-L4391

The gsm_ppp.c driver never did and does not used the L2 network interface correctly. It always called net_if_up() internally as a result from a call to `gsm_ppp_start()` I did not want to break the driver for devs using it today, who expect `gsm_ppp_start()` and `gsm_ppp_stop()` to enable the modem (which is an incorrect design according to the documentation, and is irrelevant to the L2 PPP network interface and subsystem)

Quickly, I need to apologize that I may have stated `ppp_carrier_on()/off()` is used in the `gsm_ppp.c` driver, it is not and never has been. `ppp_carrier_on()/off()` is part of the network interface subsystem, and is in use across Zephyr.

I would recommend you closely study the changes in the PR #59124 to understand the broken state it was in before, and read the linked documentation before we continue the discussion.

Reply:
Yes, I've been really trying to find a tradeoff solution all the time. See updated PR and https://github.com/zephyrproject-rtos/zephyr/pull/61147#discussion_r1286966548